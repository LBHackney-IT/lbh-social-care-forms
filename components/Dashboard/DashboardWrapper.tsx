import Head from 'next/head';
import { useRouter } from 'next/router';
import Link from 'next/link';
import s from './DashboardWrapper.module.scss';
import MyData from './MyData';

interface NavLinkProps {
  href: string;
  children: React.ReactChild;
}

const NavLink = ({ href, children }: NavLinkProps) => {
  const router = useRouter();
  return (
    <li>
      <Link href={href}>
        <a
          className={`lbh-link lbh-link--no-visited-state ${
            router.asPath === href ? s.navLinkActive : s.navLink
          }`}
        >
          {children}
        </a>
      </Link>
    </li>
  );
};

const navigation: { text: string; href: string }[] = [
  {
    text: `Allocations`,
    href: `/`,
  },
  {
    text: `Work in progress`,
    href: `/work-in-progress`,
  },
  { text: 'My work', href: '/my-work' },
];

interface Props {
  children: React.ReactChild;
}

const DashboardWrapper = ({ children }: Props): React.ReactElement => (
  <>
    <Head>
      <title>Dashboard | Social care | Hackney Council</title>
    </Head>

    <div className={`govuk-grid-row ${s.outer}`}>
      <div className="govuk-grid-column-one-quarter">
        <div className={s.sticky}>
          <nav>
            <ul className="lbh-list">
              {navigation.map((link) => (
                <NavLink href={link.href} key={link.href}>
                  {link.text}
                </NavLink>
              ))}
            </ul>
          </nav>
          <MyData />
        </div>
      </div>
      <div className="govuk-grid-column-three-quarters">{children}</div>
    </div>
  </>
);

export default DashboardWrapper;
