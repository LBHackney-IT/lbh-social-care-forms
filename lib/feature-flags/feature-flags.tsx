import { getFeatureFlags } from 'features';
import React from 'react';
import { useContext } from 'react';

export type FeatureSet = {
  [featureName: string]: {
    isActive: boolean;
  };
};

const FeatureFlagContext =
  React.createContext<FeatureSet | undefined>(undefined);

export const FeatureFlagProvider: React.FC<{
  features: FeatureSet;
}> = ({ features, children }) => {
  return (
    <FeatureFlagContext.Provider value={features}>
      {children}
    </FeatureFlagContext.Provider>
  );
};
const isFeatureActive = (features: FeatureSet) => (featureName: string) => {
  if (features[featureName] === undefined) {
    return false;
  }

  if (features[featureName].isActive === false) {
    return false;
  }

  return true;
};

/**
 * React hook for accessing feature flags
 */
export const useFeatureFlags: () => {
  isFeatureActive: (featureName: string) => boolean;
} = () => {
  const features = useContext(FeatureFlagContext);

  if (features === undefined) {
    throw new Error(
      'A <FeatureFlagProvider /> must be provided as a parent of this component'
    );
  }

  return {
    isFeatureActive: isFeatureActive(features),
  };
};

/**
 * React component for conditional rendering based on the active state of a feature flag
 */
export const ConditionalFeature: React.FC<{ name: string }> = ({
  name,
  children,
}) => {
  const { isFeatureActive } = useFeatureFlags();

  if (!isFeatureActive(name)) {
    return null;
  }

  return <>{children}</>;
};

/**
 * Typescript helper function for checking whether a given feature is active or inactive
 */
export const isFeatureFlagActive = (featureName: string) => {
  const environmentName = ['development', 'staging'].includes(
    process.env.NEXT_PUBLIC_ENV || ''
  )
    ? 'development'
    : 'production';

  const features = getFeatureFlags({ environmentName });

  return isFeatureActive(features)(featureName);
};
