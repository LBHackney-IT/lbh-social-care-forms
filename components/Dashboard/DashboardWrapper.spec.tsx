import { render } from '@testing-library/react';
import { FeatureFlagProvider } from 'lib/feature-flags/feature-flags';

import DashboardWrapper from './DashboardWrapper';

const mockedUseRouter = { pathname: '/my-cases' };

jest.mock('next/router', () => ({
  useRouter: () => mockedUseRouter,
}));

describe(`DashboardWrapper`, () => {
  it('should render properly', () => {
    const { asFragment } = render(
      <FeatureFlagProvider features={{}}>
        <DashboardWrapper>
          <div>foo</div>
        </DashboardWrapper>
      </FeatureFlagProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
