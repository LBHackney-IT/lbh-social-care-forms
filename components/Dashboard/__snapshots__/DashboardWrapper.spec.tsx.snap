// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardWrapper should render properly 1`] = `
<DocumentFragment>
  <div
    class="govuk-grid-row outer"
  >
    <div
      class="govuk-grid-column-one-quarter"
    >
      <div
        class="sticky"
      >
        <nav>
          <ul
            class="lbh-list"
          >
            <li>
              <a
                class="lbh-link lbh-link--no-visited-state navLink"
                href="/"
              >
                Allocations
              </a>
            </li>
            <li>
              <a
                class="lbh-link lbh-link--no-visited-state navLink"
                href="/work-in-progress"
              >
                Work in progress
              </a>
            </li>
            <li>
              <a
                class="lbh-link lbh-link--no-visited-state navLink"
                href="/my-work"
              >
                My work
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
    <div
      class="govuk-grid-column-three-quarters"
    >
      <div>
        foo
      </div>
    </div>
  </div>
</DocumentFragment>
`;
